internal/logger/logging_interceptor.go

```diff
@@ -101,3 +102,45 @@ func Interceptor(cfg config.LoggingConfig) grpc.UnaryServerInterceptor {
 		return resp, err
 	}
 }
+
+// RequestIDInterceptor traces request ids.
+//
+// It tries to use the request id from the request context, creating a
+// new one if that is missing. It also sends back in the trailer the
+// request id, ensuring that the client receives it.
+func RequestIDInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		rID := maybeGetRequestID(ctx)
+		ctx = zerolog.Ctx(ctx).With().Str("request_id", rID).Logger().WithContext(ctx)
+
+		resp, err := handler(ctx, req)
+
+		if err := grpc.SetTrailer(ctx, metadata.Pairs("request-id", rID)); err != nil {
+			zerolog.Ctx(ctx).Trace().Err(err).Msg("unable to attach request id to trailer")
+		}
+
+		return resp, err
+	}
+}
+
+func maybeGetRequestID(ctx context.Context) string {
+	var rID string
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		if rIDs, ok := md["request-id"]; ok {
```

Comment:
in HTTP what would this translate to? I wonder how it works with the HTTP gateway we use.

Reply:
GRPC moves back and forth headers having prefix `Grpc-Metadata-`, so it would become `Grpc-Metadata-Request-Id`.